docs/src/appendix/mast.md

```diff
@@ -0,0 +1,528 @@
+
+```rust
+pub struct DynNode;
+```
+
+The `External` node type allows us to reference `MastNode`s which are not stored in the current `MastForest` by their digest. This is what allows us to do two useful things:
+
+* Strip out parts of a program that will not be used, either because we want to keep those parts secret, or to reduce the size of the program - while preserving the integrity of the program's hash.
+* Enable a much greater degree of code reuse, and reduce the size of MAST packages. Specifically, rather than constructing a single `MastForest` representing all of the code for a given program/library and all of its dependencies; the dependencies can be referenced as external nodes, and only code unique to the program/library need be encoded in the `MastForest`. Referenc
+
+External nodes do not sacrifice any properties of the MAST, because they preserve the integrity of the MAST hash, and have the exact same semantics as if the referenced node was part of the same `MastForest`. The primary difference is that if the referenced node is not available at runtime for some reason, then any attempt to execute that node will fail.
+
+### Binary Format
+
+We can now cover the binary format for the structure we've outlined above. The two are independent of one another, but are designed such that serializing/deserializing between them is simple and efficient. First, we will cover the specification for the binary format itself, and then we will cover the specific serialization/deserialization procedures for the in-memory structure we've defined.
+
+#### Overview
+
+To make the specification easier to understand at a glance, we're going to use Rust code to define the layout, with the following modified syntax/semantics:
+
+* All structs are assumed to use `#[repr(C)]` unless a different `#[repr]` is given, or modifications are otherwise noted
+* We relax the restriction on dynamically-sized fields in the middle of structs,
+as otherwise it would not be possible to express the format in Rust syntax
+* We allow specifying the size of array-typed fields using self-referential expressions, as long as the referenced fields in the expression precede the array-typed field. For example, `items: [T; self.len]` is valid if the `len` field preceds the `items` field in the struct definition.
+
+#### Variable-Width Integers
+
+First, let's introduce a type that we will use throughout the rest of this spec, `LEB128`. It represents a variable-width, unsigned, little-endian integer. It is encoded as one or more bytes, where each byte has 7 bits to represent the value, with the last bit used to signal when another byte is needed to represent the encoded value. In this case, unless otherwise noted, we treat it as an error if an `LEB128`-encoded integer would not fit in a `usize`.
```

Comment:
For variable int encoding I've used `vint64` format in the past (see [here](https://github.com/facebook/winterfell/blob/main/utils/core/src/serde/byte_writer.rs#L78) and [here](https://github.com/facebook/winterfell/blob/main/utils/core/src/serde/byte_reader.rs#L117)). I think it is better than `LEB128`, but also it is not as widely used.

Reply:
Sure, `vint64` seems like a solid choice for our application compared to LEB128 (which is better I think for truly variable-length integer applications, but I don't think that applies to us here). I mostly chose LEB128 as a sound default choice, but I'm not too picky on what scheme we use, as long as we choose one that has an efficient encoding. I do like `vint64` for this thought!

Reply:
I've replaced the encoding scheme with `vint64` as suggested.